File: app/src/main/java/com/nhatminh/example/architecture/demoarchitecture/search/viewmodel/SearchViewModel.java
package com.nhatminh.example.architecture.demoarchitecture.search.viewmodel;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.nhatminh.example.architecture.demoarchitecture.model.GithubRepos;
import com.nhatminh.example.architecture.demoarchitecture.repository.DataRepository;

import java.util.List;

public class SearchViewModel {

    private DataRepository repository;

    private MutableLiveData<List<GithubRepos>> reposListLiveData;
    private MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);

    private MutableLiveData<DataRepository.Error> errorMessage = new MutableLiveData<>();

    public SearchViewModel(DataRepository repository) {
        this.repository = repository;
        reposListLiveData = new MutableLiveData<>();
    }

    public MutableLiveData<List<GithubRepos>> getReposListLiveData() {
        return reposListLiveData;
    }

    public MutableLiveData<DataRepository.Error> getErrorMessage() {
        return errorMessage;
    }

    public MutableLiveData<Boolean> isLoading() {
        return isLoading;
    }

    public void searchGithubRepos(String query) {
        isLoading.setValue(true);

        repository.searchRepos(query, new DataRepository.GithubDataRepositoryCallback() {
            @Override
            public void onSuccess(List<GithubRepos> reposList) {
                reposListLiveData.setValue(reposList);
                isLoading.setValue(false);
            }

            @Override
            public void onError(DataRepository.Error error) {
                errorMessage.setValue(error);
                isLoading.setValue(false);
            }
        });
    }
}

File: app/src/main/java/com/nhatminh/example/architecture/demoarchitecture/search/viewmodel/SearchViewModel.kt
package com.nhatminh.example.architecture.demoarchitecture.search.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.nhatminh.example.architecture.demoarchitecture.model.ReposData
import com.nhatminh.example.architecture.demoarchitecture.repository.DataRepository

class SearchViewModel (private val repository: DataRepository) : ViewModel() {

    private val queryStr = MutableLiveData<String>()

    var reposData : LiveData<ReposData> = Transformations.switchMap(queryStr){
        query ->
        val dataFromRepos = repository.searchRepos(query)

        Transformations.switchMap(dataFromRepos){
            reposData -> transformData(reposData)
        }
    }

    fun searchGithubRepos(query : String){
        queryStr.value = query
    }

    private fun transformData(reposData : ReposData) : LiveData<ReposData>{
        val list = reposData.reposList.filter { it.stargazersCount > 1000 }.sortedByDescending { it.stargazersCount }.toMutableList()
        return MutableLiveData<ReposData>(ReposData(list, reposData.state))
    }
}